File: 76.java
import java.util.*;

class Solution {
    public String minWindow(String s, String t) {
        int m = s.length(), n = t.length();
        if (n > m) {
            return "";
        }
        Map<Character, Integer> mapT = new HashMap<>();
        for (char i : t.toCharArray()) {
            mapT.put(i, mapT.getOrDefault(i, 0) + 1);
        }

        int have = 0, needed = mapT.size();
        int maxLength = Integer.MAX_VALUE;
        Map<Character, Integer> window = new HashMap<>();
        int left = 0, bestLeft = 0;
        for (int right = 0; right < m; right++) {
            char currChar = s.charAt(right);
            window.put(currChar, window.getOrDefault(currChar, 0) + 1);
            if (mapT.containsKey(currChar) && mapT.get(currChar).equals(window.get(currChar))) {
                have++;
            }

            while (have == needed) {
                if ((right - left) + 1 < maxLength) {
                    maxLength = (right - left) + 1;
                    bestLeft = left;
                }
                char leftChar = s.charAt(left);
                window.put(leftChar, window.get(leftChar) - 1);
                if (mapT.containsKey(leftChar) && window.get(leftChar) < mapT.get(leftChar)) {
                    have--;
                }
                left++;
            }
        }

        return maxLength == Integer.MAX_VALUE ? "" : s.substring(bestLeft, bestLeft + maxLength);
    }
}

File: Python/76.py
class Solution:
    def minWindow(self, s: str, t: str) -> str:
        if t == "":
            return ""

        count, window = {}, {}
        for c in t:
            count[c] = 1 + count.get(c, 0)

        have, need = 0, len(count)
        res, res_len = [-1, -1], float("infinity")
        left = 0
        for right in range(len(s)):
            c = s[right]
            window[c] = 1 + window.get(c, 0)

            if c in count and window[c] == count[c]:
                have += 1

            while have == need:
                if (right - left + 1) < res_len:
                    res, res_len = [left, right], (right - left + 1)
                window[s[left]] -= 1
                if s[left] in count and window[s[left]] < count[s[left]]:
                    have -= 1
                left += 1
        return s[res[0] : res[1] + 1] if res_len != float("infinity") else ""
